src/mlpack/tests/knn_test.cpp

```diff
@@ -889,6 +889,123 @@ BOOST_AUTO_TEST_CASE(DualBallTreeTest)
 }
 
 /**
+ * Test the spill tree hybrid sp-tree search (defeatist search on overlapping
+ * nodes, and backtracking in non-overlapping nodes) against the naive method.
+ * This uses only a random reference dataset.
+ */
+BOOST_AUTO_TEST_CASE(HybridSpillSearchTest)
+{
+  arma::mat dataset;
+  dataset.randu(50, 300); // 50 dimensional, 300 points.
+
+  const size_t k = 3;
+
+  KNN naive(dataset);
+  arma::Mat<size_t> neighborsNaive;
+  arma::mat distancesNaive;
+  naive.Search(dataset, k, neighborsNaive, distancesNaive);
+
+  double max_dist = 0;
+  for (size_t i = 0; i < neighborsNaive.n_cols; ++i)
+    if (distancesNaive(k - 1, i) > max_dist)
+      max_dist = distancesNaive(k - 1, i);
+
+  // If we are sure that tau is a valid strict upper bound of the kth nearest
+  // neighbor of the query points, then we can be sure that we will get an exact
+  // solution.
+  SpillSearch<> spTreeSearch(dataset, false, true,
+      max_dist * 1.01 /* tau parameter */);
+
+  for (size_t mode = 0; mode < 2; mode++)
+  {
+    switch (mode)
+    {
+      case 0: // Single Tree Search.
+        spTreeSearch.Naive() = false;
+        spTreeSearch.SingleMode() = true;
+        break;
+      case 1: // Dual Tree Search.
+        spTreeSearch.Naive() = false;
+        spTreeSearch.SingleMode() = false;
+        break;
+    }
+
+    arma::Mat<size_t> neighborsSPTree;
+    arma::mat distancesSPTree;
+    spTreeSearch.Search(dataset, k, neighborsSPTree, distancesSPTree);
+
+    for (size_t i = 0; i < neighborsSPTree.n_elem; ++i)
+    {
+      BOOST_REQUIRE_EQUAL(neighborsSPTree(i), neighborsNaive(i));
+      BOOST_REQUIRE_CLOSE(distancesSPTree(i), distancesNaive(i), 1e-5);
+    }
+  }
+}
+
+/**
+ * Test hybrid sp-tree search doesn't repeat points.
+ * This uses only a random reference dataset.
+ */
+BOOST_AUTO_TEST_CASE(DuplicatedSpillSearchTest)
+{
+  arma::mat dataset;
+  dataset.randu(50, 300); // 50 dimensional, 300 points.
+
+  const size_t k = 15;
+  double tau = 0;
+
+  for (size_t test = 0; test < 2; test++)
+  {
+    switch (test)
+    {
+      case 0:
+        tau = 0;
+        break;
+      case 1:
+        tau = 0.1;
+        break;
+    }
+
+    SpillSearch<> spTreeSearch(dataset, false, false, tau);
+    arma::Mat<size_t> neighborsSPTree;
+    arma::mat distancesSPTree;
+
+    for (size_t mode = 0; mode < 2; mode++)
+    {
+      switch (mode)
+      {
+        case 0: // Single Tree Search.
+          spTreeSearch.Naive() = false;
+          spTreeSearch.SingleMode() = true;
+          break;
+        case 1: // Dual Tree Search.
+          spTreeSearch.Naive() = false;
+          spTreeSearch.SingleMode() = false;
+          break;
+      }
```

Comment:
Done in: https://github.com/mlpack/mlpack/pull/747/commits/471bb24dc2d96620e27a7542cbba8dc8dae7ca29